### ISSUE TITLE
Cifar-10 example not converging when image_dim_ordering == 'tf'

### ISSUE THREAD

**Issue by michalgregor** [fchollet/keras#3959]
For some reason, the cifar-10 example is not converging when image_dim_ordering is set to tf. I have tried both with theano and with tensorflow – neither version works. When I switch image_dim_ordering to th all works as expected.

I have checked the data and it seems it gets loaded correctly – is it perhaps some quirk of the model or a bug in one of the layers' implementation?

**Comment by robertomest** [comment]
If you are referring to https://github.com/fchollet/keras/blob/master/examples/cifar10_cnn.py, I just tested it using data augmentation and it seems to work fine (ran up it up to 10 epochs without issue).
Some of the output:
```
Using real-time data augmentation.
Epoch 1/200
50000/50000 [==============================] - 27s - loss: 2.2058 - acc: 0.1573 - val_loss: 1.9498 - val_acc: 0.2782
Epoch 2/200
50000/50000 [==============================] - 24s - loss: 1.7706 - acc: 0.3368 - val_loss: 1.4774 - val_acc: 0.4613
Epoch 3/200
50000/50000 [==============================] - 24s - loss: 1.4741 - acc: 0.4629 - val_loss: 1.2354 - val_acc: 0.5595
Epoch 4/200
50000/50000 [==============================] - 24s - loss: 1.3342 - acc: 0.5190 - val_loss: 1.1356 - val_acc: 0.5985
Epoch 5/200
50000/50000 [==============================] - 24s - loss: 1.2240 - acc: 0.5604 - val_loss: 0.9912 - val_acc: 0.6454
Epoch 6/200
50000/50000 [==============================] - 24s - loss: 1.1415 - acc: 0.5944 - val_loss: 0.9371 - val_acc: 0.6706
Epoch 7/200
50000/50000 [==============================] - 25s - loss: 1.0796 - acc: 0.6177 - val_loss: 0.9240 - val_acc: 0.6764
Epoch 8/200
50000/50000 [==============================] - 26s - loss: 1.0293 - acc: 0.6367 - val_loss: 0.8730 - val_acc: 0.6931
Epoch 9/200
50000/50000 [==============================] - 25s - loss: 0.9979 - acc: 0.6483 - val_loss: 0.8241 - val_acc: 0.7107
Epoch 10/200
50000/50000 [==============================] - 25s - loss: 0.9691 - acc: 0.6586 - val_loss: 0.7846 - val_acc: 0.7237
```
Training without data augmentation, on the other hand, indeed failed to converge.
```
Train on 50000 samples, validate on 10000 samples
Epoch 1/200
50000/50000 [==============================] - 24s - loss: 2.3036 - acc: 0.1000 - val_loss: 2.3031 - val_acc: 0.1000
Epoch 2/200
50000/50000 [==============================] - 24s - loss: 2.3035 - acc: 0.0997 - val_loss: 2.3028 - val_acc: 0.1000
Epoch 3/200
50000/50000 [==============================] - 24s - loss: 2.3033 - acc: 0.0992 - val_loss: 2.3031 - val_acc: 0.1000
Epoch 4/200
50000/50000 [==============================] - 24s - loss: 2.3032 - acc: 0.1007 - val_loss: 2.3031 - val_acc: 0.1000
Epoch 5/200
50000/50000 [==============================] - 24s - loss: 2.3035 - acc: 0.0978 - val_loss: 2.3030 - val_acc: 0.1000
Epoch 6/200
50000/50000 [==============================] - 25s - loss: 2.3033 - acc: 0.0982 - val_loss: 2.3036 - val_acc: 0.1000
Epoch 7/200
50000/50000 [==============================] - 24s - loss: 2.3035 - acc: 0.1004 - val_loss: 2.3031 - val_acc: 0.1000
Epoch 8/200
50000/50000 [==============================] - 24s - loss: 2.3033 - acc: 0.0992 - val_loss: 2.3034 - val_acc: 0.1000
Epoch 9/200
50000/50000 [==============================] - 24s - loss: 2.3033 - acc: 0.0983 - val_loss: 2.3031 - val_acc: 0.1000
Epoch 10/200
50000/50000 [==============================] - 23s - loss: 2.3033 - acc: 0.1014 - val_loss: 2.3032 - val_acc: 0.1000
```
Halving the learning rate seems to make it work though
`sgd = SGD(lr=0.005, decay=1e-6, momentum=0.9, nesterov=True)`
Here is the output:
```
Train on 50000 samples, validate on 10000 samples
Epoch 1/200
50000/50000 [==============================] - 23s - loss: 2.1969 - acc: 0.1602 - val_loss: 1.9509 - val_acc: 0.2893
Epoch 2/200
50000/50000 [==============================] - 23s - loss: 1.8309 - acc: 0.3226 - val_loss: 1.5908 - val_acc: 0.4150
Epoch 3/200
50000/50000 [==============================] - 23s - loss: 1.5357 - acc: 0.4365 - val_loss: 1.3379 - val_acc: 0.5164
Epoch 4/200
50000/50000 [==============================] - 23s - loss: 1.3193 - acc: 0.5258 - val_loss: 1.1658 - val_acc: 0.5791
Epoch 5/200
50000/50000 [==============================] - 23s - loss: 1.1856 - acc: 0.5778 - val_loss: 1.0265 - val_acc: 0.6368
Epoch 6/200
50000/50000 [==============================] - 23s - loss: 1.0866 - acc: 0.6145 - val_loss: 0.9659 - val_acc: 0.6626
Epoch 7/200
50000/50000 [==============================] - 23s - loss: 1.0085 - acc: 0.6428 - val_loss: 0.9099 - val_acc: 0.6820
Epoch 8/200
50000/50000 [==============================] - 23s - loss: 0.9347 - acc: 0.6724 - val_loss: 0.8704 - val_acc: 0.6951
Epoch 9/200
50000/50000 [==============================] - 23s - loss: 0.8786 - acc: 0.6903 - val_loss: 0.7961 - val_acc: 0.7202
Epoch 10/200
50000/50000 [==============================] - 23s - loss: 0.8298 - acc: 0.7082 - val_loss: 0.7889 - val_acc: 0.7255
```
I'm not sure why the image_dim_ordering would make a difference.

**Comment by crowsonkb** [comment]
It doesn't look like this script seeds the RNG. Does it still exhibit this behavior if you call np.random.seed(some_int) before all Keras imports? Otherwise it could just be due to random differences in the initial weights.

**Comment by zolia** [comment]
I experience this too. After adding dim_ordering='th' to Convolution2D of cifar10_cnn.py  example, it started to converge.

**Comment by michalgregor** [comment]
OK, first of all, the issue is not augmentation-related. There is no convergence whether using augmentation or not. It also works the other way around: the "th" version converges whether using augmentation or not.

I have also tried the idea with RNG seeding. It does not really resolve the issue, but I do observe a change. E.g. with seed 74 and not using image augmentation the loss is pretty much frozen as usual at first, but then, after about 25000 steps it slowly starts decreasing. Note that this is not the correct behaviour – the error should decrease steadily from the start. With image augmentation on, even this limited amount of success is not achieved and the error stays frozen.

For reference, I am currently testing with the CPU version of tensorflow 0.11.0rc2 and keras master [https://github.com/fchollet/keras/commit/71494ffdbc8193ece7a82496ec7835f195fb8309].

I can confirm that supplying dim_ordering='th' to Convolution2D does help to work around the issue. We could, then, perhaps narrow our attention down to this type of layer.

**Comment by zachmoshe** [comment]
+1
Can confirm on MaxOSX 10.11.6, keras 1.1.0, tensorflow 0.10.0, and on Amazon g2.2xlarge, keras 1.1.1, tensorflow 0.9.0.

I build the same model and train on CIFAR10. Tested for one epoch, multiple times (from scratch). If image_dim_ordering is set to 'tf' network doesn't converge (get accuracy of ~10% which is pure random in this case). when switching to 'th' I get ~50% after first epoch.
Doesn't matter if I use CPU or GPU (on the AWS machine).

**Comment by NYcleaner** [comment]
I meet the  same question
      Can confirm on ubuntu16,04,Keras (1.1.0),tensorflow (0.11.0rc0) without GPU and another on ubuntu16,04 GPU gtx965m，cuda8.0,  keras 1.1.1,tensorflow (0.11.0rc0).， both  don't converge (get accuracy of ~10% ), Both use tf backend.
     Another interesting thing is that when I run cifar10 example  different times, sometimes it just converge and get the right result ,  but most times it doesn't converge,  val_loss result of each epoch is 0.1, so  why ?

**Comment by NYcleaner** [comment]
https://github.com/fchollet/keras/issues/4066   this link  may  solve  the question , using Adam instead of SGD.  I just try 5 times, each is  converge  with adams optimizer

**Comment by NYcleaner** [comment]
@michalgregor

**Comment by michalgregor** [comment]
That is indeed good news, thank you for pointing it out!

However, I naturally still think that it is very necessary to identify the bug that causes this behaviour because otherwise we can never be certain where it will pop up next. Essentially, I am not comfortable using the tensorflow backend for anything CNN-related at the moment, because I can never tell afterwards whether any inferior results I obtain are not due to the bug...

**Comment by NYcleaner** [comment]
yeah, what you said is really the same feeling  I  get with keras, otherwise I find the user define  callback class  may doesn't work , such like 
`class EarlyStoppingByLossVal(Callback):
    def __init__(self, monitor='val_loss', value=0.25, verbose=2):
        super(Callback, self).__init__()
        self.monitor = monitor
        self.value = value
        self.verbose = verbose

    def on_epoch_end(self, epoch, logs={}):
        current = logs.get(self.monitor)
        if current is None:
            warnings.warn("Early stopping requires %s available!" % self.monitor, RuntimeWarning)

        if current < self.value:
            if self.verbose > 0:
                print("Epoch %05d: early stopping THR" % epoch)
            self.model.stop_training = True`

I use the code class as the  model.fit  parameter， but when the val_loss reach the  bound， the whole model doesn't stop by the callback parameter I defined above
`result = model.fit(X_train, Y_train,
                  batch_size=batch_size,
                  nb_epoch=nb_epoch,
                  verbose=1,
                  validation_data=(X_test, Y_test),
                  callbacks=[EarlyStoppingByLossVal(monitor='val_loss', value=0.15, verbose=2)],
                  shuffle=True)`
That‘s a really  magic thing

**Comment by NYcleaner** [comment]
yeah, what you said is really the same feeling  I  get with keras, otherwise I find the user define  callback class  may doesn't work , such like 
`class EarlyStoppingByLossVal(Callback):
    def __init__(self, monitor='val_loss', value=0.25, verbose=2):
        super(Callback, self).__init__()
        self.monitor = monitor
        self.value = value
        self.verbose = verbose
    def on_epoch_end(self, epoch, logs={}):
        current = logs.get(self.monitor)
        if current is None:
            warnings.warn("Early stopping requires %s available!" % self.monitor, RuntimeWarning)
        if current < self.value:
            if self.verbose > 0:
                print("Epoch %05d: early stopping THR" % epoch)
            self.model.stop_training = True`

I use the code class as the  model.fit  parameter， but when the val_loss reach the  bound， the whole model doesn't stop by the callback parameter I defined above:
`result = model.fit(X_train, Y_train,
                  batch_size=batch_size,
                  nb_epoch=nb_epoch,
                  verbose=1,
                  validation_data=(X_test, Y_test),
                  callbacks=[EarlyStoppingByLossVal(monitor='val_loss', value=0.15, verbose=2)],
                  shuffle=True)`

That‘s a really  magic thing

**Comment by michalgregor** [comment]
I believe the issue is fixed with the current version of keras (1.2.1; testing with master: f41d5f021acdf29726c05f32abfbcca78edcdf21) and tensorflow (0.12.1).

If anyone still experiences it, we can reopen this.

**Comment by jrao1** [comment]
Just tried this using tf 0.12.1 and latest keras, it seems rmsprop still has trouble converging, switch to adam does solve the issue.

**Comment by michalgregor** [comment]
Well, I cannot reproduce that behaviour at all. I was testing with the original version of the cifar example, which uses SGD by default. That works. I have also tested just now with rmsprop like 20 times and it didn't freeze once. I wonder what could the difference be in your case...

**Comment by bb4242** [comment]
Hmm, I think this still may be an issue.  I just noticed the problem (rmsprop has trouble converging; switching to adam works) on keras 1.2.2 and tensorflow 1.0.0.  I'm on ubuntu 16.10.

**Comment by raginisharma14** [comment]
My loss is decreasing, but I see the accuracy always at 10%. I tried RmSProp, Adam, SGD with various learning rate values.  Is there anything which I am missing on?

**Comment by gijswobben** [comment]
I have the same issue. Simple MNIST example from the Keras website, clean install on Ubuntu 14.04, accuracy at 10% after 50 epochs..

Anyone with a working solution (I've tried every solution above)?
